### PR TITLE
[1.4] Generic HasBuff

### PR DESCRIPTION
**Pull request by ExterminatorX99**
### What is the new feature?
Generic overload for `Player.HasBuff` and `NPC.HasBuff`

### Why should this be part of tModLoader?
Reduces bloat caused by `HasBuff(ModContent.BuffType<ModBuff>())` calls

### Are there alternative designs?
Can't think of any

### Sample usage for the new feature
`player.HasBuff<ExamplePetBuff>()`

### ExampleMod updates

Used the new overload in `ExampleLightPetProjectile`, `ExamplePetProjectile` and `ExampleSimpleMinion`

### CODE REVIEW

**Comment by Steviegt6** [comment]
This is real nice, though you could just make an extension method. No reason for it not to be added unless someone wants to be a stickler for patches.

**Comment by JavidPack** [comment]
I'm not sure this is a good idea, we want to teach people that the XType is used exactly the same as the XID class, and I feel that making separate approaches like this prevents new modders from understanding that concept. Plus, if we do this for all methods that take an ID, it's going to get ugly. Not doing it for all would be just as confusing.

**Comment by Steviegt6** [comment]
> I'm not sure this is a good idea, we want to teach people that the XType is used exactly the same as the XID class, and I feel that making separate approaches like this prevents new modders from understanding that concept. Plus, if we do this for all methods that take an ID, it's going to get ugly. Not doing it for all would be just as confusing.

Extension methods it is, then.

**Review by Mirsario — APPROVED** [review]

**Comment by direwolf420** [comment]
I'd keep the method but just not use it in EM, to appease jopos point
Similar thing exists with the Recipe.AddIngredient method

File: patches/tModLoader/Terraria/NPC.TML.cs [hunk 1]
```diff
@@ -20,6 +20,10 @@ public partial class NPC : IEntityWithGlobals<GlobalNPC>
 		/// </summary>
 		public IBigProgressBar BossBar { get; set; }
 
+		public bool HasBuff(int type) => FindBuffIndex(type) != -1;
+
+		public bool HasBuff<T>() where T : ModBuff => HasBuff(ModContent.BuffType<T>());
+
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/NPC.TML.cs:26]:
```suggestion
		/// <summary> Returns whether or not this NPC currently has a (de)buff of the provided type. </summary>
		public bool HasBuff(int type) => FindBuffIndex(type) != -1;

		/// <inheritdoc cref="HasBuff(int)" />
		public bool HasBuff<T>() where T : ModBuff
			=> HasBuff(ModContent.BuffType<T>());

```

File: patches/tModLoader/Terraria/Player.TML.cs [hunk 2]
```diff
@@ -50,6 +50,10 @@ public partial class Player
 		}
 		*/
 
+		public bool HasBuff(int type) => FindBuffIndex(type) != -1;
+
+		public bool HasBuff<T>() where T : ModBuff => HasBuff(ModContent.BuffType<T>());
+
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/Player.TML.cs:56]:
```suggestion
		/// <summary> Returns whether or not this Player currently has a (de)buff of the provided type. </summary>
		public bool HasBuff(int type) => FindBuffIndex(type) != -1;

		/// <inheritdoc cref="HasBuff(int)" />
		public bool HasBuff<T>() where T : ModBuff
			=> HasBuff(ModContent.BuffType<T>());

```

**Review by Mirsario — APPROVED** [review]

**Pull request merged**